### ISSUE TITLE
mitmproxy saves failed CONNECT request URLs to HAR files incorrectly

### ISSUE THREAD

**Issue by DaniElectra** [mitmproxy/mitmproxy#6579]
#### Problem Description

mitmproxy saves failed CONNECT dumps to HAR files without having the protocol on the `url` field:

```
            {
                "startedDateTime": "2024-01-04T11:50:38.052258+00:00",
                "time": 0.37479400634765625,
                "request": {
                    "method": "CONNECT",
                    "url": "dhgyhuim.pretendo.cc:443",
                    "httpVersion": "HTTP/1.0",
                    "cookies": [],
                    "headers": [
                        {
                            "name": "User-Agent",
                            "value": "Opera/9.50 (Nintendo DSi; Opera/507; U; es-ES)"
                        },
                        {
                            "name": "Host",
                            "value": "dhgyhuim.pretendo.cc:443"
                        }
                    ],
                    "queryString": [],
                    "headersSize": 115,
                    "bodySize": 0
                },
                "response": {
                    "status": 502,
                    "statusText": "Bad Gateway",
                    "httpVersion": "HTTP/1.1",
                    "cookies": [],
                    "headers": [
                        {
                            "name": "content-length",
                            "value": "216"
                        }
                    ],
                    "content": {
                        "size": 216,
                        "compression": 0,
                        "mimeType": "",
                        "text": "Cannot connect to dhgyhuim.pretendo.cc:443: [Errno -2] Name or service not known If you plan to redirect requests away from this server, consider setting `connection_strategy` to `lazy` to suppress early connections."
                    },
                    "redirectURL": "",
                    "headersSize": 36,
                    "bodySize": 216
                },
                "cache": {},
                "timings": {
                    "connect": -1.0,
                    "ssl": -1.0,
                    "send": 0,
                    "receive": 0.37479400634765625,
                    "wait": 0
                }
            },
```

This causes visual parsing issues on other HAR parsers such as Firefox or harviewer, and in mitmproxy shows a URL parsing error, and won't try to parse any following dumps on the file:

```
[12:42:38.792] Uncaught exception POST /flows/dump (127.0.0.1)
HTTPServerRequest(protocol='http', host='127.0.0.1:8081', method='POST', uri='/flows/dump', version='HTTP/1.1', remote_ip='127.0.0.1')
Traceback (most recent call last):
  File "/home/daniel/Sources/mitmproxy/mitmproxy/io/io.py", line 53, in stream
    yield request_to_flow(request_json)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/daniel/Sources/mitmproxy/mitmproxy/io/har.py", line 80, in request_to_flow
    new_flow.request = http.Request.make(
                       ^^^^^^^^^^^^^^^^^^
  File "/home/daniel/Sources/mitmproxy/mitmproxy/http.py", line 623, in make
    req.url = url
    ^^^^^^^
  File "/home/daniel/Sources/mitmproxy/mitmproxy/http.py", line 799, in url
    self.scheme, self.host, self.port, self.path = url.parse(val)
                                                   ^^^^^^^^^^^^^^
  File "/home/daniel/Sources/mitmproxy/mitmproxy/net/http/url.py", line 55, in parse
    raise ValueError("No hostname given")
ValueError: No hostname given

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/daniel/Sources/mitmproxy/env/lib/python3.11/site-packages/tornado/web.py", line 1790, in _execute
    result = await result
             ^^^^^^^^^^^^
  File "/home/daniel/Sources/mitmproxy/mitmproxy/tools/web/app.py", line 352, in post
    for f in io.FlowReader(bio).stream():
  File "/home/daniel/Sources/mitmproxy/mitmproxy/io/io.py", line 56, in stream
    raise exceptions.FlowReadException(
mitmproxy.exceptions.FlowReadException: Unable to read HAR file. Please provide a valid HAR file
```

#### Steps to reproduce the behavior:
1. Launch mitmproxy and specify an output file on the `hardump` option.
2. Try to make a request to a domain that doesn't exist
3. Save the HAR dump and try to open the file again with mitmproxy

#### System Information

Mitmproxy: 11.0.0.dev
Python:    3.11.6
OpenSSL:   OpenSSL 3.1.4 24 Oct 2023
Platform:  Linux-6.6.8-zen1-1-zen-x86_64-with-glibc2.38

**Comment by mhils** [comment]
Could you provide an example dump file that exhibits this behavior?

**Comment by DaniElectra** [comment]
The mitmproxy flows don't seem to save this request, but here is an example HAR dump:

[test.har.zip](https://github.com/mitmproxy/mitmproxy/files/13830121/test.har.zip)

**Comment by basedBaba** [comment]
Hello! I am a first-time contributor to mitmproxy and new to open source in general, so I apologize in advance if something I say does not make sense.

My approach to this issue is to check if the request method is `CONNECT`, in which case the `url` field can be set to `https://host/` (since this behavior is seen only in case of HTTPS and not HTTP) instead of `host:port`, which is the default `url` of the CONNECT method.

Also, does the `url` field have to be of the form `https://host/` or does `https://host:port/` also work?

Please let me know if this approach make sense, so I can open a pull request for the same. Thank you!

**Comment by DaniElectra** [comment]
@basedBaba I think your approach makes sense! I can test your PR if you make one to check if it works on my end

> Also, does the `url` field have to be of the form `https://host/` or does `https://host:port/` also work?

The HAR spec says that the url field is the "absolute URL" of the request, so I would go with the `https://host:port/` form. We can also have more information about the request this way

**Comment by basedBaba** [comment]
@DaniElectra Thank you for the assistance! I have opened a PR for review.

**Comment by DaniElectra** [comment]
Fixed by #6599 :+1:
